index.rst

```diff
@@ -13,17 +13,256 @@
+
+``Butler.get``
+""""""""""""""
+
+No journaling is needed, as this is a read-only operation.
+
+``Butler.unstore`` and ``Butler.purge``
+"""""""""""""""""""""""""""""""""""""""
+
+The journal file should be written before ``Registry`` transaction is committed and deleted only after all Datastore deletions succeed.
+This is slightly problematic for client/server, because the journal file will need to be populated with information we get from the Registry database; this means the client cannot be responsible for creating the journal file unless we make fetching the datastore records and deleting them separate operations.
+That isn't too bad - it's just a slight increase in latency and a bit more http traffic.
+Another alternative would be to have the server take responsibility for creating the journal file, and then either returning responsibility for its deletion to the client or taking responsibility for both the deletion of the Datastore artifacts and the deletion of the journal file.
+Which of these is preferable probably depends on whether we want these operations to block until completion and whether we have reasons to perform other Datastore operations on the server (up to this point, having the client use a vanilla ``FileDatastore`` and perform all Datastore operations still seems viable).
+
+.. note::
+   TODO: think about how all of this interacts with DECam raws.
+
+.. note::
+   TODO: think about how all of this affects disassembled composites.
+   Or get rid of them!  I am pretty confident now that we'll never need them, and dropping that before we embark on other major changes to Datastore seems wise (sorry, Tim).
+
+Public interface changes
+========================
+
+.. note::
+   It's all stubs and outlines from here on.
+
+Bundling DatasetRef with Datastore records
+------------------------------------------
+
+- If we're passing around DatasetRefs and bundles of datastore records together a lot, we should have a class for that.
+- This would logically be what a QuantumGraph stores.
+- This + a datastore would be what backs a DeferredDatasetHandle, if we don't want that hitting a butler server or Registry unnecessarily.
+- This might replace FileLocation.
+
+Butler methods vs. Butler.registry methods
+------------------------------------------
+
+- I've totally come around to the idea of Butler being the only provider of public interfaces, and we should move all useful Registry methods there, at the same time we define a richer set of ABCs for various aspects of Butler interfaces.
+- Registry would become a nonpolymorphic concrete class used by one Butler subclass and the Butler server.
+- A lot of Registry's convenience functionality should move to Butler.
+- I think Registry's caching should move to Butler.
+- Registry's default collections and governor data ID values should definitely move to Butler.
+- Some of the convenience logic in the butler CLI scripts should also move to Butler.
```

Comment:
I'm fine with this. It's design work but be careful because the API for the "script" layer is always simplified objects (strings, numbers etc) and not objects that people really want to be using.

Reply:
I completely agree.  It's probably the display-results and confirmation-checking side of the scripts that need Python interfaces, not the "I'll run queryDatasets for you" bits.